### ISSUE TITLE
smash vs. amd

### ISSUE THREAD

**Issue by sheppard** [mbostock/smash#12]
Hi @mbostock, @jasondavies and others:

TL;DR: modules cool, but why smash when [AMD exists](http://requirejs.org/docs/whyamd.html)?

I'm excited to see that d3 modules now explicitly declare their internal dependencies!  In fact, I was half considering doing the work myself (in part so I could better understand how d3 works internally).  That said, I must confess that I'm a bit puzzled as to the purpose of creating yet-another module dependency resolver?  Don't get me wrong- I think there is often a benefit to implementing things yourself even if they have been done elsewhere.  However, Smash feels an awful lot like [RequireJS](http://requirejs.org)/AMD and [r.js](http://requirejs.org/docs/optimization.html), without some key features provided by the latter.

Specifically, in both AMD and Smash:
 - modules declare their dependencies via a simple string syntax.
 - a command-line tool (e.g. `smash` or `r.js`) resolves dependencies into path names and generates a single output file. Redundant imports are avoided.
 - said tool runs on node and can be used with either make or grunt

AMD provides a couple key advantages:
 - individual module files can be loaded as-is, on-demand, directly from the browser (great for debugging)
 - support by existing libraries like jQuery and Dojo

As I see it, the main advantage of Smash (as far as d3 is concerned) is that imported modules are executed in the same scope as their caller.  AMD modules have their own internal scope by definition, and must therefore explicitly return any exports, which are mapped to a variable in the calling script.  Thus, converting d3 source to AMD would be a much more challenging task, as the current module files all assume that the others have been executed in the same scope.  However, I would argue that in the long-term, the scoping pattern assumed by Smash will prove to be much more fragile.

The other advantage of Smash's syntax is that it is simple, and matches what the future [ECMAScript/Harmony **import** keyword](http://wiki.ecmascript.org/doku.php?id=harmony:modules) will probably look like, in contrast to the comparatively messy AMD syntax.  However, Harmony modules will probably feel much more like AMD modules (or Python modules to be precise) in that they will have their own scope with explicit exports.   This is a potential concern as Smash's use of the `import` keyword implies Harmony module compatibility when it isn't really doing the same thing.  (See http://requirejs.org/docs/whyamd.html for a further discussion of how AMD addresses potential future module patterns.)

You may legitimately ask why I am concerned how d3 modules are managed internally, if I'm only using it as a dependency.  Basically, my ultimate ideal as a mobile web developer is not exactly to create a custom d3 build (as cool as that will be).  Instead, what I would like is to create a single built JS file containing all of my internal code *and* any external modules - including specific submodules of d3 as needed.  This means that using a consistent dependency syntax for both internal and external modules is critical to making my build work.  Sure, I could make various AMD builds of d3 for each of my projects (#9), but it sure would be nice to just reference the d3 source directly.  As I see AMD slowly becoming a de-facto standard, I'm currently resorting to maintaining AMD-ified versions of popular libraries - see [wq/wq.app](http://github.com/wq/wq.app) (and it's [lib folder](https://github.com/wq/wq.app/tree/master/js/lib#third-party-libraries)) for my work to date.

My goal in posting this issue is both to better understand the reasoning behind Smash and (if you would like) to work together towards building something more compatible with RequireJS.  Let me know if there is a better place to have this discussion.

**Comment by sheppard** [comment]
Closing this in light of your work on https://github.com/d3/d3-bundler - exciting stuff!

**Comment by mbostock** [comment]
:boom:
